cfme/generic_objects/definition/ui.py

```diff
@@ -0,0 +1,466 @@
+        custom_image_file_path: generic object definition custom image file path
+        cancel: by default will not be canceled, pass True to make successful cancel
+
+    Returns:
+        GenericObjectDefinition entity
+
+    """
+    view = navigate_to(self, 'Add')
+    view.fill({
+        'name': name,
+        'description': description
+    })
+    view.associations.fill(associations)
+    view.attributes.fill(attributes)
+    view.methods.fill(methods)
+    if custom_image_file_path:
+        view.custom_image_file.file.fill(custom_image_file_path)
+        view.custom_image_file.upload_chosen_file.click()
+    if cancel:
+        view.cancel.click()
+    else:
+        view.add.click()
+        view.flash.assert_no_error()
+
+    entity = self.instantiate(
+        name=name, description=description, attributes=attributes, associations=associations,
+        methods=methods, custom_image_file_path=custom_image_file_path
+    )
+    return entity
+
+
+@MiqImplementationContext.external_for(GenericObjectDefinition.update, ViaUI)
+def update(self, updates, reset=False, cancel=False):
+    """Update generic object definition
+
+    Args:
+        updates: dict, with fields to update
+    """
+    view = navigate_to(self, 'Edit')
+    view.fill({
+        'name': updates.get('name'),
+        'description': updates.get('description')
+    })
+    view.associations.fill(updates.get('associations'))
+    view.attributes.fill(updates.get('attributes'))
+    view.methods.fill(updates.get('methods'))
+    if updates.get('custom_image_file_path'):
+        view.custom_image_file.file.fill('custom_image_file_path')
+        view.custom_image_file.upload_chosen_file.click()
+    if reset:
+        view.reset.click()
+    if cancel:
+        view.cancel.click()
+    elif not reset and not cancel:
+        view.save.click()
+        view = self.create_view(GenericObjectDefinitionDetailsView, override=updates)
+        view.flash.assert_no_error()
+
+
+@MiqImplementationContext.external_for(GenericObjectDefinition.delete, ViaUI)
+def delete(self):
+    """Delete generic object definition
+    """
+    view = navigate_to(self, 'Details')
+
+    view.configuration.item_select(
+        'Remove this Generic Object Classes from Inventory', handle_alert=True)
+    view = self.create_view(GenericObjectDefinitionAllView)
+    view.flash.assert_no_error()
+    assert view.is_displayed
+
+
+@MiqImplementationContext.external_for(GenericObjectDefinition.add_button, ViaUI)
+def add_button(self, name, description, image, request, button_type='Default', display=True,
+               dialog=None, open_url=None, display_for=None, submit_version=None,
+               system_message=None, attributes=None, role=None, button_group=None, cancel=False):
+    """Add button to generic object definition
+    """
+    if button_group:
+        view = navigate_to(self, 'AddButton', button_group=button_group)
```

Comment:
**QUESTION**
I'm not sure that passing argument into `navigate_to` it's a common thing. @psav

Reply:
Its not common, but there is at least one other instance of this type of behavior in the 'AllForProvider' nav destinations for infra virtual_machines.

I would advocate for using it only when really necessary, and when other methods won't suffice for some corner case.

In this case it looks like there is no ButtonGroup class with a 'details' view, and the navigation to a button group details page requires passing the button group name as a string. Which means `button_group` here is just a string as well, and we can't navigate to its details then call `configuration.item_select` from there.

@rlbabyuk at the least I'd like to see a TODO here for a buttongroup class with proper navigation based on `self.obj` attributes, and passing of button_group objects, not names, for these methods.